docs/pattern-matching.md

```diff
@@ -1,229 +1,366 @@
-  | GoodResult(string)
-  | NoResult;
+switch (input) {
+| Pattern1 => code1
+| Pattern2 => code2
+| Pattern3 => code3
+};
 ```
 
-While using the `switch` expression on it, you can "destructure" it:
+The switch finds the first pattern that matches the input, and then executes the code that the pattern points to (the code after the `=>`). Code for a pattern can be a single expression, or a block of statements.
 
 ```reason
-let data = GoodResult("Product shipped!");
-
-let message =
-  switch (data) {
-  | GoodResult(theMessage) => "Success! " ++ theMessage
-  | BadResult(errorCode) => "Something's wrong. The error code is: " ++ string_of_int(errorCode)
+type t = A | B | C;
+
+let y =
+  switch (x) {
+  | A => "zero"
+  | B =>
+    f();
+    "one";
+  | C => "two"
   };
 ```
 
-Notice how we've destructured `data` while handling each different case. The above `switch` will give you a compiler warning:
+If `x` were `B`, then the second pattern is matched, causing `f(); "one";` to be executed, which causes `y` to be set to `"one"`. Note that each code block in the switch must evaluate to the same type (`string` in this example).
 
-```
-Warning 8: this pattern-matching is not exhaustive.
-Here is an example of a value that is not matched:
-NoResult
-```
+## Patterns
 
-Isn't that great? While matching on the shape of your data, the type system warned of an unhandled case. This **conditional** aspect is what makes it pattern matching rather than plain destructuring. Most data structures with an "**if this then that**" aspect work with it:
+### Primitives
+
+The simplest patterns are primitive values, which are checked for equality.
 
 ```reason
-switch (myList) {
-| [] => print_endline("Empty list")
-| [a, ...theRest] => print_endline("list with the head value " ++ a)
+switch (x) {
+| true => f("t")
+| false => g("f")
 };
 
-switch (myArray) {
-| [|1, 2|] => print_endline("This is an array with item 1 and 2")
-| [||] => print_endline("This array has no element")
-| _ => print_endline("This is an array")
+switch (x) {
+| "a" => 4.0
+| "tree" => 1.23
+| _ => 77.5
 };
 ```
 
-The `_` case is a special fall-through case that allows all unmatched conditions to go to that branch.
+### Variables
+
+Variables can be created from patterns. In the previous examples, the `_` variable acted as a catch-all, matching all remaining values (see [Exhaustive warning](#exhaustive-warning)). You could instead create a variable without a leading underscore to use it later in the block.
+
+```reason
+switch (f()) {
+| 0 => "zero"
+| 1 => "one"
+| k => "another number " ++ string_of_int(k)
+};
+```
 
-You can even switch on string, int and others. You can even have many patterns going to the same result!
+Note that if a variable with the same name already exists in the scope of the switch, then it is not used in the pattern, and is instead shadowed by the pattern's variable. Variables in patterns are declarations of new variables, not references to existing ones.
```

Comment:
> , then it is not used in the pattern, 

"then it will be shadowed by a new variable with that same name, throughout the statements to the right hand side of the next `=>`. That original value that was already inside a variable by the same name is not used a matching constraint for the pattern match. Variable names in pattern matches always create new variables in scope.